### ISSUE TITLE
Waavi\Translation\TranslationServiceProvider:register() being called multiple times.

### ISSUE THREAD

**Issue by cirdog** [Waavi/translation#38]
First of all thanks for your thoughtful package, it's the most comprehensive solution I've found on the web so far.

But if you are aware of my reply on another issue, I've somehow resolved the "ReflectionException" problem caused by a malformed bootstrap/cache/services.php. And during my time with the debugger, I've discovered that register() of TranslationServiceProvider was being called multiple times, apparently doing the exact same thing 2 - 3 times over, and I think this could lead to some performance issues.

Firstly before I've added that publishes() method to the ServiceProvider, I've got 2 repeated calls on register(), and after I've added that, I've got 3 calls.

I've actually skimmed though the call stack from the debugger, the subsequent repeated calls were invoked by the "Validator" (via some getDependencies() method) as I was tracing it during the "out of the box" authentication. 

I'm not 100% sure about this, but it looks like some other providers depend on the original "translator" service (the service name being used everywhere I believe), and when they invoke that, since it now points to your provider, ~~it re-does everything from start to end (translator, translation.loader, translation.cache.repositories), so this may explain why it's 3 calls (wild guess).~~

~~So is there any official way to avoid this redundant instantiation? I've also noticed within the Laravel Container, somewhere it checks for already-instantiated instances, could it be a "missed registration" that causes Laravel to re-instantiate your service provider over again? I could be all wrong so please forgive me guesses.~~

~~Sorry I was busy on stuff so I haven't yet gotten to the bottom of the stack trace, but I hope to contribute some input on this matter. Thanks for your inspiring package again.~~

I'll do some more stack tracing when free.

**Comment by cirdog** [comment]
**UPDATE (2016-04-27):**

I've traced it through and now know what was happening.

Your **LanguageRepository**'s __constructor() is different from your **TranslationRepository** in superficially-minor way, but it triggers a rather major consequence.

```
public function __construct(Language $model, Validator $validator, Config $config)
    {
        $this->model                   = $model;
        $this->validator               = $validator;
        $this->defaultLocale           = $config->get('app.locale');
        $this->defaultAvailableLocales = $config->get('translator.available_locales', []);
        $this->config                  = $config;
    }
```

```
public function validate(array $attributes)
    {
        $id    = array_get($attributes, 'id', 'NULL');
        $table = $this->model->getTable();
        $rules = [
            'locale' => "required|unique:{$table},locale,{$id}",
            'name'   => "required|unique:{$table},name,{$id}",
        ];
        $validator = $this->validator->make($attributes, $rules);
        if ($validator->fails()) {
            $this->errors = $validator->errors();
            return false;
        }
        return true;
    }
```

The $this->validator relies on a Validator injection from the constructor, but unfortunately ValidatorServiceProvider has a callback in it that injects TranslatorServiceProvider. And since your package takes over the 'translator' service, it causes some kind of recursion here. In the end your package is being instantiated repeatedly before the first call-stack is even finished, therefore the "loadedProviders" cannot suppress the repeated instantiations.

So the solution is simple, just make it use the $this->app['validator'] singleton instead of re-injecting it at construction, then all repeated calls are gone and I've shaved off 100ms on average during my casual test run using Debugbar.

Thanks again for your great package, are you free to edit this or you want a pull-request? I hope you come back more often because I appreciate your sharing very much!!

**Comment by sildraug** [comment]
Thank you so very much for your very insightful work!!!

Sorry I didn't reply to you any sooner, just had my first kid and have been really busy for the past two months. Your fix has been applied on version 2.1.4.2.

Please let me know if you have any more suggestions, critiques, etc...
